master/jobmaster/benchmark/master.go

```diff
@@ -0,0 +1,275 @@
+package benchmark
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/hanfei1991/microcosom/master/cluster"
+	"github.com/hanfei1991/microcosom/model"
+	"github.com/hanfei1991/microcosom/pb"
+	"github.com/hanfei1991/microcosom/pkg/log"
+	"go.uber.org/zap"
+)
+
+// Master implements the master of benchmark workload.
+type Master struct {
+	*Config
+	job *model.Job
+	ctx context.Context
+
+	resouceManager cluster.ResourceMgr
+	client         cluster.ExecutorClient
+
+	offExecutors chan model.ExecutorID
+
+	mu           sync.Mutex
+	execTasks    map[model.ExecutorID][]*model.Task
+	runningTasks map[model.TaskID]*Task
+
+	scheduleWaitingTasks chan scheduleGroup
+}
+
+// scheduleGroup is the min unit of scheduler, and the tasks in the same group have to be scheduled in the same node.
+type scheduleGroup []*Task
+
+// TaskStatus represents the current status of the task.
+type TaskStatus int32
+
+const (
+	// Running means the task is running.
+	Running TaskStatus = iota
+	// Stopped means the task has been stopped by any means.
+	Stopped
+	// Finished means the task has finished its job.
+	Finished
+)
+
+// Task is the container of a dispatched tasks,  and records its status.
+type Task struct {
+	*model.Task
+
+	exec   model.ExecutorID
+	status TaskStatus
+}
+
+// ID implements JobMaster interface.
+func (m *Master) ID() model.JobID {
+	return m.job.ID
+}
+
+// master dispatches a set of task.
+func (m *Master) dispatch(tasks []*Task) error {
+	arrangement := make(map[model.ExecutorID][]*model.Task)
+	for _, task := range tasks {
+		subjob, ok := arrangement[task.exec]
+		if !ok {
+			arrangement[task.exec] = []*model.Task{task.Task}
+		} else {
+			subjob = append(subjob, task.Task)
+			arrangement[task.exec] = subjob
```

Comment:
Does a `SubJob` mean a list of tasks belonging to a specific executor?

Reply:
yep